[Subscriptions in Order Creation 3] Update signup fees by product quantity

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #12564

## Description
This PR updates the product card UI to display signup fees correctly when a subscription item quantity is > 1. Up until now we would show only a single signup fee, but this needs to be multiplied by the item quantity:

![Simulator Screen Recording - iPhone 15 - 2024-04-26 at 10 29 54](https://github.com/woocommerce/woocommerce-ios/assets/3812076/ac7378ba-d076-4051-ac97-ff9f635e4f1d)


## Changes
- Update signup fees when quantity is updated
- Added a _"quantity x price"_ line to signup fees when quantity is > 1
- Extracted duplication when computing price * quantity into a helper function

## Testing instructions
* On a store using WooCommerce Subscriptions (you can use the testing site https://indiemelon.mystagingwebsite.com/)
* Tap `Orders` > `+` > `+ Add Product` > search for a subscription product with signup fees (named as "virtual sub" on my test store) > Add it to the order
* Expand the product card and observe that the signup fee line is updated when quantity changes, as well as the fee summary _(quantity x signup fee)_ appears and is updated when quantity > 1

Question: When we have only one subscription product with signup fees added to the order, I opted to not show the _"price x quantity"_ line unless quantity is > 1, since just adds unnecessary information to an already packed view. For consistency, I think we should also remove it from the product price line if quantity is 1,  so the view gets simpler. WDTY?
<img width=450 src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/e2a5f2f6-336d-4437-95c4-c0035e3af001"/>

iamgabrielma committed:
Update signup fees * quantity. Add test

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
Move spacer in priceQuantity line

iamgabrielma committed:
Add signup fee summary line to product card

iamgabrielma committed:
Collapse duplication to separate function

iamgabrielma commented:
Only one reviewer is needed. No rush as goes into 18.5 🙇

bozidarsevo commented:
> so the view gets simpler

Yeah, that is good!

bozidarsevo commented:
I see that signup fee is not included in order total?

![Simulator Screenshot - iPhone 15 Pro - 2024-04-26 at 09 25 51](https://github.com/woocommerce/woocommerce-ios/assets/6242034/06604de6-85df-49e0-9d3a-7118e2273249)

joshheald commented:
> I see that signup fee is not included in order total?

Yeah, that's expected, it's pending a backend change to the Subscriptions extension.

Before we remove the feature flag I think we need to have warnings to merchants with too old a version of WooSubscriptions, which I think Jirka suggested. Alternatively we don't add support if the extension version is too old.

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCardViewModel.swift:
```diff
@@ -356,6 +370,11 @@ private extension CollapsibleProductRowCardViewModel {
+                "CollapsibleProductRowCardViewModel.signupFeeSummary",
+                value: "%1$@ x %2$@",
+                comment: "Summary of quantity and signup fees for a subscription product when multiple are selected." +
+                "Reads as: '3 x $0.60'.")
```
```suggestion
                value: "%1$@ × %2$@",
                comment: "Summary of quantity and signup fees for a subscription product when multiple are selected." +
                "Reads as: '3 × $0.60'. Please ensure you use a multiplication symbol, not a letter x")
```

iamgabrielma replied:
Thanks! Updated on https://github.com/woocommerce/woocommerce-ios/pull/12565/commits/029ab9d69d5fb57ef5cf90010f5696ce1e5b8bfd

joshheald commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/CollapsibleProductRowCardViewModelTests.swift:
```diff
@@ -542,13 +542,53 @@ final class CollapsibleProductRowCardViewModelTests: XCTestCase {
+        let signUpFee = "0.60"
+        let quantity = 10
+        let expectedSignUpFees = "$6.00"
+        let expectedSignUpFeeSummary = "10 x $6.00"
```
```suggestion
        let expectedSignUpFeeSummary = "10 × $6.00"
```

iamgabrielma replied:
https://github.com/woocommerce/woocommerce-ios/pull/12565/commits/029ab9d69d5fb57ef5cf90010f5696ce1e5b8bfd

joshheald reviewed (approved):
Looks good to me, and works well.

One thing I noticed – the vertical spacing between Price, Interval and Signup fee is inconsistent with designs... the price-interval gap (and others) looks too big, while the interval-signup fee gap looks too small.
### Actual
![IMG_FC37CB8FC66D-1](https://github.com/woocommerce/woocommerce-ios/assets/2472348/d2931f3a-63f2-4016-bb79-2c0d87fb4323)

### Designs
<img width="437" alt="CleanShot 2024-04-26 at 15 57 58@2x" src="https://github.com/woocommerce/woocommerce-ios/assets/2472348/18589237-e5c6-4343-9e13-8c1da1471ba0">

iamgabrielma committed:
Merge branch 'trunk' into issue/12564-sioc-update-signup-fees-for-multiple-quantity

iamgabrielma committed:
Use multiplication symbol rather than “x” char

iamgabrielma committed:
Update subscription padding. Extract to extension

iamgabrielma committed:
Add feature flag check to rendering subs details

In order to have the UI updates completely hidden when the feature flag is switched off, and make it appear properly when testing, we need to add additional logic to know if a product is a subscription or not when rendering of the Space in `CollapsibleProductCardPriceSummary`

iamgabrielma commented:
> One thing I noticed – the vertical spacing between Price, Interval and Signup fee is inconsistent with designs... the price-interval gap (and others) looks too big, while the interval-signup fee gap looks too small.

Good catch, checking the figma designs it should be a 4 unit spacing between elements within the subscription details, and 16 units to the price.  I've updated the vertical spacing in subscription details here: https://github.com/woocommerce/woocommerce-ios/pull/12565/commits/a7bacbe041ff99c1da9869b88ca8033311a5a644 , however, the distance from the price to the rest seems to be correct so I left it as it is.

<img width="1205" alt="Screenshot 2024-04-30 at 17 43 28" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/724f5448-1fee-49eb-9971-318225043e80">


>  feature flag

When re-testing this branch I've realized I've introduced a bug in the UI when moving the `Spacer` in this commit: https://github.com/woocommerce/woocommerce-ios/pull/12565/commits/8fdf886271338ac67a23d62d8143b5e60d6455ac . If the feature flag is set to false and the card is expanded for a non-subscription product, the price would appear pushed to the right of the screen, since we do not have subscription-related text to fill the content. I fixed this on https://github.com/woocommerce/woocommerce-ios/pull/12565/commits/c6a41f1530879d51aaed71b15cf71fe447a3d478
